File: php/utility/PrepareAuth.php
<?php
declare(strict_types=1);

// XenoCanto SDK utility: prepare_auth

class XenoCantoPrepareAuth
{
    private const QUERY_AUTH = 'key';
    private const OPTION_APIKEY = 'apikey';
    private const NOT_FOUND = '__NOTFOUND__';

    public static function call(XenoCantoContext $ctx): array
    {
        $spec = $ctx->spec;
        if (!$spec) {
            return [null, $ctx->make_error('auth_no_spec', 'Expected context spec property to be defined.')];
        }

        $query = &$spec->query;
        $options = $ctx->client->options_map();

        // Public APIs that need no auth omit the options.auth block entirely.
        if (!isset($options['auth']) || $options['auth'] === null) {
            unset($query[self::QUERY_AUTH]);
            return [$spec, null];
        }

        $apikey = \Voxgig\Struct\Struct::getprop($options, self::OPTION_APIKEY, self::NOT_FOUND);

        if (
            (is_string($apikey) && ($apikey === self::NOT_FOUND || $apikey === ''))
            || $apikey === null
        ) {
            unset($query[self::QUERY_AUTH]);
        } else {
            // NO PREFIX IN A QUERY STRING. `?token=Bearer%20abc` is not a
            // thing any API reads; the prefix is a header convention and is
            // dropped here deliberately rather than silently concatenated.
            $query[self::QUERY_AUTH] = is_string($apikey) ? $apikey : '';
        }

        return [$spec, null];
    }
}

File: lua/utility/prepare_auth.lua
-- XenoCanto SDK utility: prepare_auth

local vs = require("utility.struct.struct")

local QUERY_AUTH = "key"
local OPTION_APIKEY = "apikey"
local NOT_FOUND = "__NOTFOUND__"

local function prepare_auth_util(ctx)
  local spec = ctx.spec
  if spec == nil then
    return nil, ctx:make_error("auth_no_spec",
      "Expected context spec property to be defined.")
  end

  local query = spec.query
  local options = ctx.client:options_map()

  -- Public APIs that need no auth omit the options.auth block entirely.
  if options.auth == nil then
    query[QUERY_AUTH] = nil
    return spec, nil
  end

  local apikey = vs.getprop(options, OPTION_APIKEY, NOT_FOUND)

  if apikey == nil
    or (type(apikey) == "string" and (apikey == NOT_FOUND or apikey == ""))
  then
    query[QUERY_AUTH] = nil
  else
    -- NO PREFIX IN A QUERY STRING: "?token=Bearer%20abc" is not a thing
    -- any API reads, so the auth.prefix a header placement space-joins is
    -- dropped here deliberately rather than concatenated.
    local apikey_val = ""
    if type(apikey) == "string" then
      apikey_val = apikey
    end
    query[QUERY_AUTH] = apikey_val
  end

  return spec, nil
end

return prepare_auth_util
